trie/patricia.go

```diff
@@ -99,48 +112,29 @@ func (b *branch) increase(key []byte) (int, int, int) {
 	return 0, 0, 1
 }
 
-// collapse updates the node, returns the <key, value> if the node can be collapsed
-// value is the hash of only remaining leaf node, another DB access is needed to get the actual value
-func (b *branch) collapse(k, v []byte, index byte, childClps bool) ([]byte, []byte, bool) {
-	// if child cannot collapse, no need to check and return false
-	if !childClps {
-		return k, v, false
-	}
-	// value == nil means no entry exist on the incoming path, trim it
-	if v == nil {
+// collapse updates the node, returns the path and key if the node can be collapsed
+func (b *branch) collapse(key []byte, index byte) ([]byte, []byte, int, error) {
+	if key == nil {
 		b.Path[index] = nil
-	}
-	// count number of remaining path
-	nb := 0
-	var key, value []byte
-	for i := 0; i < RADIX; i++ {
-		if len(b.Path[i]) > 0 {
-			nb++
-			key = []byte{byte(i)}
-			value = b.Path[i]
+		// count number of remaining path
+		nb := 0
+		path := byte(0)
+		for i := 0; i < RADIX; i++ {
+			if len(b.Path[i]) > 0 {
+				nb++
+				key = b.Path[i]
+				path = byte(i)
+			}
 		}
+		logger.Debug().Uint8("branch", index).Msg("trim")
+		return []byte{path}, key, nb, nil
```

Comment:
yes, collapse() tries to see if this node can collapse (be removed) along the path to root

key is the hash of downstream node,  key == nil means the downstream nodes has been deleted, so first we trim this branch by setting b.Path[index] = nil, then we count how many branches are still active on this branch node, and return the corresponding key

for instance, say b only has 2 active branches: b.Path[5] = 0x93ac4568..., b.Path[7] = 0x812e334d... now b.collapse(nil, 5) is called, this means node on branch 5 is deleted, so it returns 7, 0x812e334d, 1, nil

when i saw nb=1 returned, i know there is only 1 active branch, so it can turn into a Ext node e with e.Path = 7, e.Value = 0x812e334d to make the trie simple, the branch can be deleted as well

but if b has another active b.Path[4], then it will return nb = 2, and the branch node will be kept